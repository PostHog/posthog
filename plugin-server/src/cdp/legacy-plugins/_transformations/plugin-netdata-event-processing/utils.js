export function cleanPropertyName(k) {
    return k
        // convert to lower case
        .toLowerCase()
        // remove leading slash
        .replace(/^\//, "")
        // replace all slashes and dots with _
        .replace(/\/|\.|-| /g, "_")
        ;
}

export function isStringDDMMYYYYHHMM(dt){
    var reDate = /^((0?[1-9]|[12][0-9]|3[01])[- /.](0?[1-9]|1[012])[- /.](19|20)?[0-9]{2}[ ][012][0-9][:][0-9]{2})*$/;
    return reDate.test(dt);
}

export function isDemo(url) {
    if (
        (url.includes('://london.my-netdata.io'))
        ||
        (url.includes('://london3.my-netdata.io'))
        ||
        (url.includes('://cdn77.my-netdata.io'))
        ||
        (url.includes('://octopuscs.my-netdata.io'))
        ||
        (url.includes('://bangalore.my-netdata.io'))
        ||
        (url.includes('://frankfurt.my-netdata.io'))
        ||
        (url.includes('://newyork.my-netdata.io'))
        ||
        (url.includes('://sanfrancisco.my-netdata.io'))
        ||
        (url.includes('://singapore.my-netdata.io'))
        ||
        (url.includes('://toronto.my-netdata.io'))
        ){
        return true
    } else {
        return false
    }
}

export function splitPathName(event) {
    if (event.properties['$pathname']) {
        event.properties["$pathname"].split("/").forEach((pathname, index) => {
          if ((pathname !== "") && (pathname !== null)){
            event.properties[`pathname_${index}`] = pathname
          }
        })
      }
    return event
}