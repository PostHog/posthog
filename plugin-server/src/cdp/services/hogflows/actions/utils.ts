import { HogFlow, HogFlowAction } from '~/schema/hogflow'

export const findActionById = (hogFlow: HogFlow, id: string): HogFlowAction => {
    const action = hogFlow.actions.find((action) => action.id === id)
    if (!action) {
        throw new Error(`Action ${id} not found`)
    }

    return action
}

export const findActionByType = <T extends HogFlowAction['type']>(
    hogFlow: HogFlow,
    type: T
): Extract<HogFlowAction, { type: T }> | undefined => {
    const action = hogFlow.actions.find((action) => action.type === type)
    if (!action) {
        return undefined
    }

    return action as Extract<HogFlowAction, { type: T }>
}

export const findNextAction = (
    hogFlow: HogFlow,
    currentActionId: string,
    edgeIndex?: number
): HogFlowAction | undefined => {
    const edges = hogFlow.edges.filter((edge) => edge.from === currentActionId)

    let nextActionId: string | undefined

    if (edgeIndex === undefined) {
        nextActionId = edges.find((edge) => edge.type === 'continue')?.to
    } else {
        nextActionId = edges.find((edge) => edge.type === 'branch' && edge.index === edgeIndex)?.to
    }

    return nextActionId ? findActionById(hogFlow, nextActionId) : undefined
}
