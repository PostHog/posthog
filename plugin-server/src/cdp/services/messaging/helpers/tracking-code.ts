import { CyclotronJobInvocationHogFunction } from '~/cdp/types'
import { defaultConfig } from '~/config/config'

function toBase64UrlSafe(input: string) {
    // Encode to normal base64
    const b64 = Buffer.from(input, 'utf8').toString('base64')
    // Make URL safe and strip padding
    return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

function fromBase64UrlSafe(b64url: string) {
    // Restore base64 from URL-safe variant
    let b64 = b64url.replace(/-/g, '+').replace(/_/g, '/')
    // Pad to length multiple of 4
    while (b64.length % 4) {
        b64 += '='
    }
    return Buffer.from(b64, 'base64').toString('utf8')
}

export const parseEmailTrackingCode = (
    encodedTrackingCode: string
): { functionId: string; invocationId: string } | null => {
    // customId  is like ph_fn_id=function-1&ph_inv_id=invocation-1
    const decodedTrackingCode = fromBase64UrlSafe(encodedTrackingCode)
    try {
        const [functionId, invocationId] = decodedTrackingCode.split(':')
        if (!functionId || !invocationId) {
            return null
        }
        return { functionId, invocationId }
    } catch {
        return null
    }
}

export const generateEmailTrackingCode = (
    invocation: Pick<CyclotronJobInvocationHogFunction, 'functionId' | 'id'>
): string => {
    // Generate a base64 encoded string free of equal signs
    return toBase64UrlSafe(`${invocation.functionId}:${invocation.id}`)
}

export const generateEmailTrackingPixelUrl = (
    invocation: Pick<CyclotronJobInvocationHogFunction, 'functionId' | 'id'>
): string => {
    return `${defaultConfig.CDP_EMAIL_TRACKING_URL}/public/m/pixel?ph_id=${generateEmailTrackingCode(invocation)}`
}
