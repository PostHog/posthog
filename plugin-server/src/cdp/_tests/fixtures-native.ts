import { UUIDT } from '~/utils/utils'

import { CyclotronJobInvocationHogFunction, HogFunctionType } from '../types'
import { createHogFunction } from './fixtures'
import { SAMPLE_GLOBALS } from './fixtures'

export const createExampleNativeInvocation = (
    hogFunctionOverrides: Partial<HogFunctionType> = {},
    inputs: Record<string, any> = {}
): CyclotronJobInvocationHogFunction => {
    const hogFunction = createHogFunction(hogFunctionOverrides)

    return {
        id: new UUIDT().toString(),
        state: {
            globals: {
                inputs,
                ...SAMPLE_GLOBALS,
            },
            timings: [],
            attempts: 0,
        },
        teamId: hogFunction.team_id,
        functionId: hogFunction.id,
        hogFunction,
        queue: 'native',
        queuePriority: 0,
    }
}
