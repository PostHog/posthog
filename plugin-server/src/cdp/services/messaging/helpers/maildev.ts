import nodemailer from 'nodemailer'

import { registerShutdownHandler } from '~/lifecycle'
import { isDevEnv, isTestEnv } from '~/utils/env-utils'
import { fetch } from '~/utils/request'

export const mailDevTransport =
    isDevEnv() || isTestEnv()
        ? nodemailer.createTransport({
              url: `http://${process.env.MAILDEV_HOST || '127.0.0.1'}:${process.env.MAILDEV_PORT || 1025}`,
              ignoreTLS: true,
              secure: false, // MailDev doesn't use TLS
              connectionTimeout: 1000, // ms
              greetingTimeout: 1000,
              socketTimeout: 1000,
          })
        : null

export const mailDevWebUrl = `http://${process.env.MAILDEV_HOST || '127.0.0.1'}:${process.env.MAILDEV_WEB_PORT || 1080}`

registerShutdownHandler(() => {
    return Promise.resolve(mailDevTransport?.close())
})

export class MailDevAPI {
    constructor() {}

    async getEmails(): Promise<any[]> {
        const response = await fetch(`${mailDevWebUrl}/email`)
        return response.json()
    }

    async clearEmails(): Promise<void> {
        await fetch(`${mailDevWebUrl}/email/all`, {
            method: 'DELETE',
        })
    }
}
