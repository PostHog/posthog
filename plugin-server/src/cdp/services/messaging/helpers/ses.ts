import crypto from 'node:crypto'
import { z } from 'zod'

import { MinimalAppMetric } from '~/cdp/types'
import { parseJSON } from '~/utils/json-parse'
import { logger } from '~/utils/logger'
import { fetch } from '~/utils/request'

import { parseEmailTrackingCode } from './tracking-code'

/**
 * ---------- SNS envelope types ----------
 * If raw_message_delivery=false (default), SNS wraps your message in this envelope.
 */
const SnsEnvelopeSchema = z.object({
    Type: z.enum(['SubscriptionConfirmation', 'Notification', 'UnsubscribeConfirmation']),
    MessageId: z.string(),
    Token: z.string().optional(),
    TopicArn: z.string(),
    Subject: z.string().optional(),
    Message: z.string(), // either SES event JSON (Notification) or a confirmation message
    Timestamp: z.string(),
    SignatureVersion: z.enum(['1']),
    Signature: z.string(),
    SigningCertURL: z.string().url(),
    UnsubscribeURL: z.string().url().optional(),
})

export type SnsEnvelope = z.infer<typeof SnsEnvelopeSchema>

/**
 * ---------- SES event types ----------
 * AWS posts an array of records in the "Message" (or directly as body if raw_message_delivery=true).
 * We model the common fields and create specific event detail types.
 */
const SesMailSchema = z.object({
    timestamp: z.string(),
    source: z.string(), // From address
    messageId: z.string(),
    destination: z.array(z.string()),
    headersTruncated: z.boolean().optional(),
    headers: z.array(z.object({ name: z.string(), value: z.string() })).optional(),
    tags: z.record(z.array(z.string())).optional(), // your message tags: { user_id: ["u_123"] }
})

const SesCommonEventBase = z.object({
    eventType: z.enum([
        'Send',
        'Reject',
        'Bounce',
        'Complaint',
        'Delivery',
        'Open',
        'Click',
        'RenderingFailure',
        'DeliveryDelay',
    ]),
    mail: SesMailSchema,
})

const SesOpenEventSchema = SesCommonEventBase.extend({
    eventType: z.literal('Open'),
    open: z.object({
        ipAddress: z.string().optional(),
        userAgent: z.string().optional(),
        timestamp: z.string(),
    }),
})

const SesClickEventSchema = SesCommonEventBase.extend({
    eventType: z.literal('Click'),
    click: z.object({
        ipAddress: z.string().optional(),
        link: z.string(),
        userAgent: z.string().optional(),
        timestamp: z.string(),
    }),
})

const SesDeliveryEventSchema = SesCommonEventBase.extend({
    eventType: z.literal('Delivery'),
    delivery: z.object({
        processingTimeMillis: z.number().optional(),
        smtpResponse: z.string().optional(),
        reportingMTA: z.string().optional(),
        timestamp: z.string(),
        recipients: z.array(z.string()).optional(),
    }),
})

const SesBounceEventSchema = SesCommonEventBase.extend({
    eventType: z.literal('Bounce'),
    bounce: z.object({
        bounceType: z.enum(['Undetermined', 'Permanent', 'Transient']),
        bounceSubType: z.string().optional(),
        bouncedRecipients: z.array(
            z.object({
                emailAddress: z.string(),
                action: z.string().optional(),
                status: z.string().optional(),
                diagnosticCode: z.string().optional(),
            })
        ),
        timestamp: z.string(),
        reportingMTA: z.string().optional(),
    }),
})

const SesComplaintEventSchema = SesCommonEventBase.extend({
    eventType: z.literal('Complaint'),
    complaint: z.object({
        complainedRecipients: z.array(z.object({ emailAddress: z.string() })),
        timestamp: z.string(),
        complaintFeedbackType: z.string().optional(),
        userAgent: z.string().optional(),
        feedbackId: z.string().optional(),
    }),
})

const SesRenderingFailureSchema = SesCommonEventBase.extend({
    eventType: z.literal('RenderingFailure'),
    renderingFailure: z.object({
        errorMessage: z.string(),
        templateName: z.string().optional(),
    }),
})

const SesSendEventSchema = SesCommonEventBase.extend({ eventType: z.literal('Send') })
const SesRejectEventSchema = SesCommonEventBase.extend({ eventType: z.literal('Reject') })

const SesEventRecordSchema = z.union([
    SesOpenEventSchema,
    SesClickEventSchema,
    SesDeliveryEventSchema,
    SesBounceEventSchema,
    SesComplaintEventSchema,
    SesRenderingFailureSchema,
    SesSendEventSchema,
    SesRejectEventSchema,
])

const SesEventBatchSchema = z.array(SesEventRecordSchema)

export type SesEventRecord = z.infer<typeof SesEventRecordSchema>

const EVENT_TYPE_TO_METRIC_NAME: Record<SesEventRecord['eventType'], MinimalAppMetric['metric_name']> = {
    Open: 'email_opened',
    Click: 'email_link_clicked',
    // Delivery: 'email_sent',
    Bounce: 'email_bounced',
    Complaint: 'email_blocked',
    RenderingFailure: 'email_failed',
    Send: 'email_sent',
    Reject: 'email_failed',
    Delivery: 'email_sent',
}

export class SesWebhookHandler {
    certCache: Record<string, Promise<string> | undefined> = {}

    private async fetchText(url: string): Promise<string> {
        const response = await fetch(url)
        if (response.status >= 400) {
            throw new Error(`Failed to fetch ${url}: ${response.status}`)
        }
        return await response.text()
    }

    private async fetchCert(url: string): Promise<string> {
        // Validate that the URL is from AWS SNS
        if (!this.isValidSnsCertUrl(url)) {
            throw new Error(`Invalid SNS certificate URL: ${url}`)
        }

        if (this.certCache[url]) {
            return await this.certCache[url]!
        }
        this.certCache[url] = this.fetchText(url)
        return this.certCache[url]!
    }

    private isValidSnsCertUrl(url: string): boolean {
        try {
            const parsedUrl = new URL(url)

            // Must be HTTPS
            if (parsedUrl.protocol !== 'https:') {
                return false
            }

            // Must be from sns.{region}.amazonaws.com
            const hostname = parsedUrl.hostname
            if (!hostname.match(/^sns\.[a-z0-9-]+\.amazonaws\.com$/)) {
                return false
            }

            // Must end with .pem
            if (!parsedUrl.pathname.endsWith('.pem')) {
                return false
            }

            return true
        } catch {
            return false
        }
    }

    /**
     * Parse incoming body accounting for SNS raw vs envelope
     */
    private parseIncomingBody(
        body: unknown
    ):
        | { mode: 'raw' | 'sns'; records: SesEventRecord[] }
        | { mode: 'sns'; envelope: SnsEnvelope; records?: SesEventRecord[] } {
        // If it's already an object with "Type", it's probably the SNS envelope (raw=false)
        if (body && typeof body === 'object' && 'Type' in (body as any)) {
            const env = SnsEnvelopeSchema.parse(body)
            if (env.Type === 'Notification') {
                const inner = parseJSON(env.Message)
                const record = SesEventRecordSchema.parse(inner)
                return { mode: 'sns', envelope: env, records: [record] }
            }
            // For non-Notification, return envelope; caller decides how to handle
            return { mode: 'sns', envelope: env }
        }

        // raw_message_delivery=true â†’ body is already the SES array
        const records = SesEventBatchSchema.parse(body)
        return { mode: 'raw', records }
    }

    /**
     * Verify SNS signature
     * Best practice: verify the signature unless you're behind a trusted ALB / private VPC.
     */
    private async verifySnsSignature(envelope: SnsEnvelope): Promise<boolean> {
        try {
            // 1) Fetch cert
            const cert = await this.fetchCert(envelope.SigningCertURL)
            // 2) Build string to sign (per SNS docs)
            const stringToSign = this.buildStringToSign(envelope)
            // 3) Verify
            const verifier = crypto.createVerify('RSA-SHA1') // SNS SignatureVersion=1 uses SHA1
            verifier.update(stringToSign, 'utf8')
            return verifier.verify(cert, envelope.Signature, 'base64')
        } catch {
            return false
        }
    }

    /**
     * Build string to sign for SNS signature verification
     */
    private buildStringToSign(m: SnsEnvelope): string {
        // Follows AWS SNS docs for SignatureVersion=1
        // For Notification:
        //   "Message\n{Message}\nMessageId\n{MessageId}\nSubject\n{Subject}\nTimestamp\n{Timestamp}\nTopicArn\n{TopicArn}\nType\n{Type}\n"
        // Subject line omitted if not present.
        const lines: string[] = []
        const pushKV = (k: string, v?: string) => {
            if (v !== undefined) {
                lines.push(k)
                lines.push(v)
            }
        }
        if (m.Type === 'Notification') {
            pushKV('Message', m.Message)
            pushKV('MessageId', m.MessageId)
            pushKV('Subject', m.Subject)
            pushKV('Timestamp', m.Timestamp)
            pushKV('TopicArn', m.TopicArn)
            pushKV('Type', m.Type)
        } else if (m.Type === 'SubscriptionConfirmation' || m.Type === 'UnsubscribeConfirmation') {
            pushKV('Message', m.Message)
            pushKV('MessageId', m.MessageId)
            pushKV('SubscribeURL', (m as any).SubscribeURL) // present in confirmation payload body, not in envelope schema
            pushKV('Timestamp', m.Timestamp)
            pushKV('Token', m.Token!)
            pushKV('TopicArn', m.TopicArn)
            pushKV('Type', m.Type)
        }
        return lines.join('\n') + '\n'
    }

    async handleWebhook(opts: {
        body: any
        headers: Record<string, string | string[] | undefined>
        verifySignature?: boolean
    }): Promise<{
        status: number
        body: unknown
        metrics?: {
            functionId?: string
            invocationId?: string
            metricName: MinimalAppMetric['metric_name']
        }[]
    }> {
        logger.info('[SesWebhookHandler] handleWebhook', { body: opts.body, headers: opts.headers })
        const parsed = this.parseIncomingBody(opts.body)

        logger.info('[SesWebhookHandler] parsed', { parsed })

        // If SNS envelope present and verification requested, verify signature
        if ('envelope' in parsed && opts.verifySignature) {
            logger.info('[SesWebhookHandler] verifying signature', { envelope: parsed.envelope })
            const ok = await this.verifySnsSignature(parsed.envelope)
            logger.info('[SesWebhookHandler] signature verified', { ok })
            if (!ok) {
                return { status: 403, body: { error: 'Invalid SNS signature' } }
            }
        }

        // Handle confirmation flow
        if (parsed.mode === 'sns' && 'envelope' in parsed && parsed.envelope?.Type === 'SubscriptionConfirmation') {
            logger.info('[SesWebhookHandler] confirming subscription', { envelope: parsed.envelope })
            // Confirm by visiting SubscribeURL (contained in the *message JSON*, not envelope.Message field here)
            // We need to fetch the inner message JSON to get SubscribeURL
            const env = parsed.envelope
            const inner = parseJSON(env.Message) as { SubscribeURL?: string }
            logger.info('[SesWebhookHandler] confirming subscription', { inner })
            if (inner.SubscribeURL) {
                await this.fetchText(inner.SubscribeURL)
            }
            return { status: 200, body: { ok: true } }
        }

        if (parsed.mode === 'sns' && 'envelope' in parsed && parsed.envelope?.Type === 'UnsubscribeConfirmation') {
            logger.info('[SesWebhookHandler] confirming unsubscribe', { envelope: parsed.envelope })
            return { status: 200, body: { ok: true } }
        }

        // Process SES events
        const records = parsed.mode === 'raw' ? parsed.records : parsed.records!
        const metrics: {
            functionId?: string
            invocationId?: string
            metricName: MinimalAppMetric['metric_name']
        }[] = []

        for (const rec of records) {
            logger.info('[SesWebhookHandler] processing record', { rec })
            const tags = rec.mail.tags
            const { functionId, invocationId } = parseEmailTrackingCode(tags?.ph_id?.[0] || '') || {}

            if (!functionId && !invocationId) {
                logger.error('[SesWebhookHandler] handleWebhook: No functionId or invocationId found', { rec })
                continue
            }

            const metricName = EVENT_TYPE_TO_METRIC_NAME[rec.eventType]
            metrics.push({ functionId, invocationId, metricName })
        }

        return { status: 200, body: { ok: true }, metrics }
    }
}
