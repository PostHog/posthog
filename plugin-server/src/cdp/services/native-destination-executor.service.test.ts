import { DateTime, Settings } from 'luxon'

import { defaultConfig } from '~/config/config'
import { forSnapshot } from '~/tests/helpers/snapshots'
import { fetch, FetchResponse } from '~/utils/request'

import { createHogFunction } from '../_tests/fixtures'
import { createExampleNativeInvocation } from '../_tests/fixtures-native'
import { NativeDestinationExecutorService } from './native-destination-executor.service'

const inputs = {
    url: 'https://posthog.com/webhook',
    method: 'POST',
    body: {
        event_name: '$pageview',
    },
    headers: {
        Authorization: 'Bearer abc',
    },
    debug_mode: true,
}

describe('NativeDestinationExecutorService', () => {
    let service: NativeDestinationExecutorService
    let mockFetch: jest.Mock<Promise<FetchResponse>, Parameters<typeof fetch>>

    beforeEach(() => {
        Settings.defaultZone = 'UTC'
        service = new NativeDestinationExecutorService(defaultConfig)

        service.fetch = mockFetch = jest.fn((_url, _options) =>
            Promise.resolve({
                status: 200,
                json: () => Promise.resolve({}),
                text: () => Promise.resolve(JSON.stringify({})),
                headers: {},
            } as any)
        )

        const fixedTime = DateTime.fromObject({ year: 2025, month: 1, day: 1 }, { zone: 'UTC' })
        jest.spyOn(Date, 'now').mockReturnValue(fixedTime.toMillis())
    })

    afterAll(() => {
        jest.useRealTimers()
    })

    describe('native plugins', () => {
        it('should call the plugin perform method', async () => {
            const fn = createHogFunction({
                name: 'Plugin test',
                template_id: 'native-webhook',
            })

            const invocation = createExampleNativeInvocation(fn, inputs)

            mockFetch.mockResolvedValue({
                status: 200,
                json: () => Promise.resolve({ total_count: 1 }),
                text: () =>
                    Promise.resolve(
                        JSON.stringify({
                            code: 200,
                            server_upload_time: 1747910402315,
                            payload_size_bytes: 22207,
                            events_ingested: 1,
                        })
                    ),
                headers: {},
            })

            const result = await service.execute(invocation)

            expect(result.finished).toBe(true)

            result.logs.forEach((x) => {
                if (typeof x.message === 'string' && x.message.includes('Function completed in')) {
                    x.message = 'Function completed in [REPLACED]'
                }
            })

            expect(result.logs).toMatchSnapshot()

            expect(mockFetch).toHaveBeenCalledTimes(1)
            expect(forSnapshot(mockFetch.mock.calls[0])).toMatchInlineSnapshot(`
                [
                  "https://posthog.com/webhook",
                  {
                    "body": "{"event_name":"$pageview"}",
                    "headers": {
                      "Authorization": "Bearer abc",
                      "Content-Type": "application/json",
                      "User-Agent": "PostHog.com/1.0",
                    },
                    "method": "POST",
                  },
                ]
            `)
        })

        it('should handle non retryable fetch errors', async () => {
            const fn = createHogFunction({
                name: 'Plugin test',
                template_id: 'native-webhook',
            })

            const invocation = createExampleNativeInvocation(fn, inputs)

            mockFetch.mockResolvedValue({
                status: 403,
                json: () => Promise.resolve({ error: 'Forbidden' }),
                text: () => Promise.resolve(JSON.stringify({ error: 'Forbidden' })),
                headers: { 'retry-after': '60' },
            })

            const result = await service.execute(invocation)

            expect(result.finished).toBe(true)

            result.logs.forEach((x) => {
                if (typeof x.message === 'string' && x.message.includes('Function completed in')) {
                    x.message = 'Function completed in [REPLACED]'
                }
            })

            expect(result.logs).toMatchSnapshot()

            expect(mockFetch).toHaveBeenCalledTimes(1)
            expect(forSnapshot(mockFetch.mock.calls[0])).toMatchInlineSnapshot(`
                [
                  "https://posthog.com/webhook",
                  {
                    "body": "{"event_name":"$pageview"}",
                    "headers": {
                      "Authorization": "Bearer abc",
                      "Content-Type": "application/json",
                      "User-Agent": "PostHog.com/1.0",
                    },
                    "method": "POST",
                  },
                ]
            `)

            expect(result.invocation).toEqual({
                functionId: expect.any(String),
                hogFunction: expect.any(Object),
                id: expect.any(String),
                queue: 'native',
                queueMetadata: {
                    tries: 1,
                },
                queuePriority: 0,
                queueParameters: undefined,
                queueScheduledAt: undefined,
                queueSource: undefined,
                state: expect.any(Object),
                teamId: 1,
            })
        })

        it('should retry retryable fetch errors', async () => {
            const fn = createHogFunction({
                name: 'Plugin test',
                template_id: 'native-webhook',
            })

            const invocation = createExampleNativeInvocation(fn, inputs)

            mockFetch.mockResolvedValue({
                status: 429,
                json: () => Promise.resolve({ error: 'Too many requests' }),
                text: () => Promise.resolve(JSON.stringify({ error: 'Too many requests' })),
                headers: { 'retry-after': '60' },
            })

            const result = await service.execute(invocation)

            expect(result.finished).toBe(false)

            result.logs.forEach((x) => {
                if (typeof x.message === 'string' && x.message.includes('Function completed in')) {
                    x.message = 'Function completed in [REPLACED]'
                }
            })

            expect(result.logs).toMatchSnapshot()

            expect(mockFetch).toHaveBeenCalledTimes(1)
            expect(forSnapshot(mockFetch.mock.calls[0])).toMatchInlineSnapshot(`
                [
                  "https://posthog.com/webhook",
                  {
                    "body": "{"event_name":"$pageview"}",
                    "headers": {
                      "Authorization": "Bearer abc",
                      "Content-Type": "application/json",
                      "User-Agent": "PostHog.com/1.0",
                    },
                    "method": "POST",
                  },
                ]
            `)

            expect(result.invocation).toEqual({
                hogFunction: expect.any(Object),
                functionId: expect.any(String),
                id: expect.any(String),
                queue: 'native',
                queueMetadata: { tries: 1 },
                queueParameters: undefined,
                queuePriority: 1,
                queueScheduledAt: expect.any(Object),
                queueSource: undefined,
                state: expect.any(Object),
                teamId: 1,
            })

            let minBackoffMs = DateTime.utc().plus({ milliseconds: defaultConfig.CDP_FETCH_BACKOFF_BASE_MS }).toMillis()
            let maxBackoffMs = DateTime.utc()
                .plus({ milliseconds: defaultConfig.CDP_FETCH_BACKOFF_BASE_MS * 2 })
                .toMillis()
            let scheduledAt = result.invocation.queueScheduledAt!.toMillis()
            expect(scheduledAt > minBackoffMs && scheduledAt < maxBackoffMs).toBe(true)

            // second fetch call

            const invocationResults2 = await service.execute(result.invocation)

            expect(invocationResults2.finished).toBe(false)

            expect(invocationResults2.invocation).toEqual({
                hogFunction: expect.any(Object),
                functionId: expect.any(String),
                id: expect.any(String),
                queue: 'native',
                queueMetadata: {
                    tries: 2,
                },
                queueParameters: undefined,
                queuePriority: 2,
                queueScheduledAt: expect.any(Object),
                queueSource: undefined,
                state: expect.any(Object),
                teamId: 1,
            })

            minBackoffMs = DateTime.utc()
                .plus({ milliseconds: defaultConfig.CDP_FETCH_BACKOFF_BASE_MS * 2 })
                .toMillis()
            maxBackoffMs = DateTime.utc()
                .plus({ milliseconds: defaultConfig.CDP_FETCH_BACKOFF_BASE_MS * 3 })
                .toMillis()
            scheduledAt = invocationResults2.invocation.queueScheduledAt!.toMillis()
            expect(scheduledAt > minBackoffMs && scheduledAt < maxBackoffMs).toBe(true)

            // third fetch call

            const invocationResults3 = await service.execute(invocationResults2.invocation)

            expect(invocationResults3.finished).toBe(true)

            expect(invocationResults3.invocation).toEqual({
                hogFunction: expect.any(Object),
                functionId: expect.any(String),
                id: expect.any(String),
                queue: 'native',
                queueMetadata: { tries: 3 },
                queueParameters: undefined,
                queuePriority: 0,
                queueScheduledAt: undefined,
                queueSource: undefined,
                state: expect.any(Object),
                teamId: 1,
            })
        })

        // TODO: write tests for multiple fetches and the throwHttpErrors flag
    })
})
