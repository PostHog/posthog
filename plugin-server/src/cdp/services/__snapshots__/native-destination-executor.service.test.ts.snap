// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeDestinationExecutorService native plugins should call the plugin perform method 1`] = `
[
  {
    "level": "debug",
    "message": "config, {"url":"https://posthog.com/webhook","method":"POST","body":{"event_name":"$pageview"},"headers":{"Authorization":"Bearer abc"},"debug_mode":true}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "endpoint, https://posthog.com/webhook",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "options, {"method":"POST","headers":{"Authorization":"Bearer abc"},"json":{"event_name":"$pageview"}}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "fetchOptions, {"method":"POST","headers":{"User-Agent":"PostHog.com/1.0","Authorization":"Bearer abc","Content-Type":"application/json"},"body":"{\\"event_name\\":\\"$pageview\\"}"}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "convertedResponse, 200, {"code":200,"server_upload_time":1747910402315,"payload_size_bytes":22207,"events_ingested":1}, {"code":200,"server_upload_time":1747910402315,"payload_size_bytes":22207,"events_ingested":1}, {}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "info",
    "message": "Function completed in [REPLACED]",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
]
`;

exports[`NativeDestinationExecutorService native plugins should handle non retryable fetch errors 1`] = `
[
  {
    "level": "debug",
    "message": "config, {"url":"https://posthog.com/webhook","method":"POST","body":{"event_name":"$pageview"},"headers":{"Authorization":"Bearer abc"},"debug_mode":true}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "endpoint, https://posthog.com/webhook",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "options, {"method":"POST","headers":{"Authorization":"Bearer abc"},"json":{"event_name":"$pageview"}}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "fetchOptions, {"method":"POST","headers":{"User-Agent":"PostHog.com/1.0","Authorization":"Bearer abc","Content-Type":"application/json"},"body":"{\\"event_name\\":\\"$pageview\\"}"}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "warn",
    "message": "HTTP request failed with status 403 ({"error":"Forbidden"}). ",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "error",
    "message": "Function failed: Error executing function on event uuid: Request failed with status 403 ({"error":"Forbidden"})",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
]
`;

exports[`NativeDestinationExecutorService native plugins should retry retryable fetch errors 1`] = `
[
  {
    "level": "debug",
    "message": "config, {"url":"https://posthog.com/webhook","method":"POST","body":{"event_name":"$pageview"},"headers":{"Authorization":"Bearer abc"},"debug_mode":true}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "endpoint, https://posthog.com/webhook",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "options, {"method":"POST","headers":{"Authorization":"Bearer abc"},"json":{"event_name":"$pageview"}}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "debug",
    "message": "fetchOptions, {"method":"POST","headers":{"User-Agent":"PostHog.com/1.0","Authorization":"Bearer abc","Content-Type":"application/json"},"body":"{\\"event_name\\":\\"$pageview\\"}"}",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
  {
    "level": "warn",
    "message": "HTTP request failed with status 429 ({"error":"Too many requests"}). Scheduling retry...",
    "timestamp": "2025-01-01T00:00:00.000Z",
  },
]
`;
